### PR TITLE
Slider resizing

### PR DESCRIPTION
**Pull request by reichlfl**
Sliders are now resizable by clicking a small area at the left or right edge.

Resizing the slider keeps the non-clicked edge fixed and adjust the
currently displayed data window according to the new slider size.

Both sliders support this behavior, but it is only enabled for the horizontal
slider at this point.
Missing for the vertical slider:

- Rendering of the small arrows inside the slider
- Callback implementation in CaptureWindow to react to the resizing
changes of the slider.

New scrollbar:
image

Clicking highlights only the arrow:
image

Resizing the slider resizes the visible data window:
image

This was a bit of back and forth between the unit tests, integrating the slider into 
the capture window, and rework / bugfixing, so it's not easily separated into
smaller chunks. I rebased it into 4 bigger commits, but some contain issues that
were fixed lateron. Let me know if it's too hard to follow.

### CODE REVIEW

**Review by pierricgimmig — APPROVED** [review]
Hi Florian, slider resizing is really cool! Thanks a lot. I also love the page scroll feature. I tested it everything works as expected. Please see comments, mainly regarding units and naming, but in general LGTM!

File: OrbitGl/SliderTest.cpp [hunk 1]
```diff
@@ -14,48 +14,271 @@ class MockCanvas : public GlCanvas {
...
-  MockCanvas canvas;
+template <>
+void pick<0>(GlSlider& slider, int start, int other_dim) {
+  slider.OnPick(start, other_dim);
+}
+
+template <>
+void pick<1>(GlSlider& slider, int start, int other_dim) {
+  slider.OnPick(other_dim, start);
+}
```
Comment (beckerhe) [inline OrbitGl/SliderTest.cpp]: `if constexpr(dim == 0) {` might make this cleaner. WDYT?

**Review by beckerhe — APPROVED** [review]
LGTM! Minor comments! Thanks!

File: OrbitGl/GlSlider.cpp [hunk 2]
```diff
@@ -83,99 +148,118 @@ void GlSlider::DrawSlider(GlCanvas* canvas, float x, float y, float width, float
...
+
+  DrawSlider(canvas, start, y, slider_width, GetPixelHeight(), ShadingDirection::kTopToBottom);
+
+  Batcher* batcher = canvas->GetBatcher();
+  const float kEpsilon = 0.0001f;
 
-  float start = ratio_ * nonSliderWidth;
+  // Left / right resize arrows and separator
+  const float height_factor = 2.f;
+  float tri_size = slider_resize_pixel_margin_ - 3.f;
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.cpp:224]: The slider has a 2px border, and since the available space for the arrow is quite limited, I've decided to have the tip of the arrow touch this border and leave 1px between the vertical line and the arrow. It's a bit arbitrary, but IMHO it looks better than a super-small arrow.

I'll add a comment to explain that.

File: OrbitGl/GlSlider.cpp [hunk 3]
```diff
@@ -48,6 +48,35 @@ Color GlSlider::GetDarkerColor(const Color& color) {
...
+
+  ratio_ = clamp(ratio, 0.f, 1.f);
+
+  if (drag_callback_) {
+    drag_callback_(ratio_);
+  }
+}
+
+void GlSlider::OnPick(int x, int y) {
+  int value = GetRelevantMouseDim(x, y);
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.cpp]: True, it's not super easy to understand. Changed that, made the GlSlider constructor protected to make sure that is_vertical_ is only set once upon construction, and only be the deriving classes.

File: OrbitGl/CaptureWindow.cpp [hunk 4]
```diff
@@ -571,11 +572,16 @@ void CaptureWindow::OnVerticalDrag(float ratio) {
   NeedsUpdate();
 }
 
+void CaptureWindow::OnZoom(float start, float end) {
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/CaptureWindow.cpp:575]: Done

File: OrbitGl/GlSlider.cpp [hunk 5]
```diff
@@ -13,25 +15,28 @@ const float GlSlider::kGradientFactor = 0.25f;
...
-void GlSlider::SetSliderRatio(float ratio)  // [0,1]
+void GlSlider::SetSliderPosRatio(float start_ratio)  // [0,1]
 {
-  ratio_ = ratio;
+  pos_ratio_ = clamp(start_ratio, 0.f, 1.f);
+  right_edge_ratio_ = pos_ratio_ * (1.0f - length_ratio_) + length_ratio_;
 }
 
-void GlSlider::SetSliderWidthRatio(float width_radio)  // [0,1]
+void GlSlider::SetSliderLengthRatio(float length_ratio)  // [0,1]
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.cpp]: Yes! Normalized is the much better word that I was looking for to replace ratio :D

File: OrbitGl/GlSlider.cpp [hunk 6]
```diff
@@ -83,99 +148,118 @@ void GlSlider::DrawSlider(GlCanvas* canvas, float x, float y, float width, float
...
-  float newX = static_cast<float>(x) - sliderLeftWidth;
-  float ratio = newX / nonSliderWidth;
-
-  ratio_ = clamp(ratio, 0.f, 1.f);
-
-  if (drag_callback_) {
-    drag_callback_(ratio_);
-  }
+float GlVerticalSlider::GetCanvasEdgeLength() {
+  return canvas_->GetHeight() - GetOrthogonalSliderSize();
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.cpp]: I've renamed it to `GetBarPixelLength()` now. The wording currently should be consistent with "slider" for the actual pickable slider, and "bar" for the background. As far as I can see there's no widely accepted convention... I've seen it called "slider", "thumb", "tab", "knob"...

File: OrbitGl/GlSlider.cpp [hunk 7]
```diff
@@ -83,99 +148,118 @@ void GlSlider::DrawSlider(GlCanvas* canvas, float x, float y, float width, float
...
-
+void GlHorizontalSlider::Draw(GlCanvas* canvas, PickingMode /*picking_mode*/) {
   canvas_ = canvas;
   static float y = 0;
 
-  float canvasWidth = canvas_->GetWidth() - GetOrthogonalSliderSize();
-  float sliderWidth = length_ * canvasWidth;
-  float nonSliderWidth = canvasWidth - sliderWidth;
+  float canvas_width = canvas_->GetWidth() - GetOrthogonalSliderSize();
+  float slider_width = ceilf(length_ratio_ * canvas_width);
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.cpp:208]: See comment above

File: OrbitGl/GlSlider.cpp [hunk 8]
```diff
@@ -83,99 +148,118 @@ void GlSlider::DrawSlider(GlCanvas* canvas, float x, float y, float width, float
...
+  float non_slider_width = canvas_width - slider_width;
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.cpp]: And above

File: OrbitGl/GlSlider.cpp [hunk 9]
```diff
@@ -83,99 +148,118 @@ void GlSlider::DrawSlider(GlCanvas* canvas, float x, float y, float width, float
...
+  float start = floorf(pos_ratio_ * non_slider_width);
+
+  DrawSlider(canvas, start, y, slider_width, GetPixelHeight(), ShadingDirection::kTopToBottom);
+
+  Batcher* batcher = canvas->GetBatcher();
+  const float kEpsilon = 0.0001f;
 
-  float start = ratio_ * nonSliderWidth;
+  // Left / right resize arrows and separator
+  const float height_factor = 2.f;
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.cpp]: Done

File: OrbitGl/GlSlider.h [hunk 10]
```diff
@@ -34,13 +34,23 @@ class GlSlider : public Pickable, public std::enable_shared_from_this<GlSlider>
   typedef std::function<void(float)> DragCallback;
   void SetDragCallback(DragCallback callback) { drag_callback_ = callback; }
 
+  enum class ResizeDirection { kUpdateMin, kUpdateMax };
+  typedef std::function<void(float, ResizeDirection)> ResizeCallback;
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.h]: Thanks, yes, that's better

File: OrbitGl/GlSlider.h [hunk 11]
```diff
@@ -34,13 +34,23 @@ class GlSlider : public Pickable, public std::enable_shared_from_this<GlSlider>
...
+  void SetResizeCallback(ResizeCallback callback) { resize_callback_ = callback; }
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.h]: Done

File: OrbitGl/SliderTest.cpp [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  hslider.SetDragCallback([&](float ratio) { hpos = ratio; });
+
+  hslider.OnPick(0, 0);
+  hslider.OnDrag(1000, 0);
+
+  // Expect the slider to be dragged all the way to the right
+  // (overshoot, first, then go back to exact drag pos)
+  EXPECT_NEAR(hpos, 1.f, kEpsilon);
+  hslider.OnDrag(500, 0);
+  EXPECT_NEAR(hpos, 1.0f, kEpsilon);
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/SliderTest.cpp]: Yes, I was trying to remove the .0f everywhere, must have slipped, thanks.

File: OrbitGl/GlSlider.h [hunk 13]
```diff
@@ -34,10 +34,12 @@ class GlSlider : public Pickable, public std::enable_shared_from_this<GlSlider>
   typedef std::function<void(float)> DragCallback;
   void SetDragCallback(DragCallback callback) { drag_callback_ = callback; }
 
-  enum class ResizeDirection { kUpdateMin, kUpdateMax };
-  typedef std::function<void(float, ResizeDirection)> ResizeCallback;
+  typedef std::function<void(float)> ResizeCallback;
   void SetResizeCallback(ResizeCallback callback) { resize_callback_ = callback; }
 
+  [[nodiscard]] float GetPosRatio() { return pos_ratio_; }
+  [[nodiscard]] float GetLengthRatio() { return length_ratio_; }
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.h]: These and some additional methods :-)

File: OrbitGl/GlSlider.h [hunk 14]
```diff
@@ -52,21 +54,37 @@ class GlSlider : public Pickable, public std::enable_shared_from_this<GlSlider>
 
   [[nodiscard]] virtual float GetCanvasEdgeLength() = 0;
 
+  [[nodiscard]] float PixelToLen(float value) { return value / GetCanvasEdgeLength(); }
+  [[nodiscard]] float LenToPixel(float value) { return value * GetCanvasEdgeLength(); }
+  [[nodiscard]] float PixelToPos(float value) { return value / LenToPixel(1.0f - length_ratio_); }
+  [[nodiscard]] float PosToPixel(float value) { return value * LenToPixel(1.0f - length_ratio_); }
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.h]: Indeed

File: OrbitGl/SliderTest.cpp (hunk 1 above)
Comment (reichlfl) [reply OrbitGl/SliderTest.cpp]: Good point, thanks for the suggestion.

File: OrbitGl/GlSlider.cpp [hunk 15]
```diff
@@ -13,25 +13,28 @@ const float GlSlider::kGradientFactor = 0.25f;
...
       bar_color_(61, 61, 61, 255),
       min_slider_pixel_length_(20),
+      slider_resize_pixel_margin_(5),
       pixel_height_(20),
       orthogonal_slider_size_(20) {}
 
 void GlSlider::SetSliderPosRatio(float start_ratio)  // [0,1]
 {
-  ratio_ = start_ratio;
+  pos_ratio_ = clamp(start_ratio, 0.f, 1.f);
```
[the remark this answers is not in the document]
Comment (reichlfl) [reply OrbitGl/GlSlider.cpp]: Done
